project-management/src/main/java/life/qbic/projectmanagement/application/measurement/MeasurementService.java

```diff
@@ -582,78 +454,126 @@ private List<MeasurementId> performRegistrationPxp(
-
-    if (sampleIdCodeEntries.size() != associatedSampleCodes.size()) {
-      log.error(
-          "Could not find all corresponding sample ids for input: " + associatedSampleCodes);
-      throw new MeasurementRegistrationException(ErrorCode.FAILED);
+  /**
+   * Merges and builds {@link ProteomicsMeasurement} based on the given pool information.
+   *
+   * @param groupedMetadata already grouped measurement metadata by pool
+   * @param projectId       the project the measurement belongs to
+   * @return
+   * @since 1.0.0
+   */
+  private Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> mergeByPool(
+      Map<String, List<ProteomicsMeasurementMetadata>> groupedMetadata, ProjectId projectId) {
+    Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> metadataMap = new HashMap<>();
+    for (String poolLabel : groupedMetadata.keySet()) {
+      var measurements = groupedMetadata.get(poolLabel);
+      metadataMap.putAll(build(measurements, projectId));
     }
+    return metadataMap;
+  }
 
-    var instrumentQuery = resolveOntologyCURI(metadata.instrumentCURI());
-    if (instrumentQuery.isEmpty()) {
-      throw new MeasurementRegistrationException(ErrorCode.UNKNOWN_ONTOLOGY_TERM);
+  private Map<SampleCode, SampleIdCodeEntry> buildSampleIdLookupTable(
+      Collection<ProteomicsMeasurementMetadata> metadata) {
+    Map<SampleCode, SampleIdCodeEntry> sampleIdLookupTable = new HashMap<>();
+    var sampleCodes = metadata.stream().map(ProteomicsMeasurementMetadata::sampleCode).toList();
+    for (SampleCode sampleCode : sampleCodes) {
+      var sampleIdQueryResult = queryIdCodePair(sampleCode).orElseThrow();
+      sampleIdLookupTable.put(sampleCode, sampleIdQueryResult);
     }
+    return sampleIdLookupTable;
+  }
+
+  /**
+   * Builds an instance of {@link ProteomicsMeasurement} and its corresponding
+   * {@link SampleIdCodeEntry}s based on the given list of {@link ProteomicsMethodMetadata} and the
+   * {@link ProjectId}.
+   * <p>
+   * Disclaimer: the method does NOT evaluate, of the entries belong together, e.g. they are part of
+   * the same pool. This grouping needs to be done by the caller.
+   *
+   * @param metadataList A list of pre-grouped metadata of a pool or single measurement representing
+   *                     one unit of a {@link ProteomicsMeasurement}.
+   * @param projectId    the project the measurement belongs to
+   * @return
+   * @since 1.0.0
+   */
+  private Map<ProteomicsMeasurement, Collection<SampleIdCodeEntry>> build(
+      List<ProteomicsMeasurementMetadata> metadataList, ProjectId projectId) {
+    Map<SampleCode, SampleIdCodeEntry> sampleIdLookupTable = buildSampleIdLookupTable(metadataList);
+    var sampleCodes = sampleIdLookupTable.keySet();
+    var specificMetadata = createSpecificMetadata(metadataList, sampleIdLookupTable);
+    var assignedMeasurementCode = MeasurementCode.createMS(sampleCodes.iterator().next().code());
```

Comment:
Accoding to the JavaDoc of the Set#iterator():
> Returns an iterator over the elements in this set. The elements are returned in no particular order (unless this set is an instance of some class that provides a guarantee)

I am unsure what the assigned measurementCode needs in regards of sample codes and it might not matter which of the sample codes it gets. But be aware that this is the behaviour. You can leave it as is if this is what you expect.

Reply:
which sample code is taken in a pool is irrelevant, since the semantic encoding is already broken due to the 1:n relationship of measurement and sample it derives. 
A solution would be to refactor the measurement code completely and not include the sample code at all. Out of scope of this PR though

Reply:
thank you for the clarification.